### PR TITLE
Add DiplomatOption support to C and CPP backends

### PR DESCRIPTION
**Pull request by Manishearth**
Progress on #246

<s>Based on #558</s>

### CODE REVIEW

File: tool/templates/c/enum.h.jinja [hunk 1]
```diff
@@ -5,4 +5,4 @@ enum {{ ty_name }} {
   {%- endfor %}
 } {%- if !is_for_cpp %} {{ ty_name }} {%- endif %};
 
-
+typedef struct {{ ty_name }}_option {union { {{ty_name}} ok; }; bool is_ok; } {{ ty_name }}_option;
```
Comment (sffc) [inline tool/templates/c/enum.h.jinja:8]: Question: why the `union`? To avoid calling the default constructor?

File: feature_tests/cpp/include/OptionInputStruct.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    
+    
+    } // extern "C"
+} // namespace capi
+} // namespace
+
+
+inline diplomat::capi::OptionInputStruct OptionInputStruct::AsFFI() const {
+  return diplomat::capi::OptionInputStruct {
+    /* .a = */ a.has_value() ? (diplomat::capi::OptionU8{ { a.value() }, true }) : (diplomat::capi::OptionU8{ {}, false }),
```
Comment (sffc) [inline feature_tests/cpp/include/OptionInputStruct.hpp:28]: Question/Suggestion: why is this code inlined instead of invoking `OptionU8::AsFFI` ?

File: feature_tests/cpp/include/OptionInputStruct.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    /* .b = */ b.has_value() ? (diplomat::capi::OptionChar{ { b.value() }, true }) : (diplomat::capi::OptionChar{ {}, false }),
+    /* .c = */ c.has_value() ? (diplomat::capi::OptionEnum_option{ { c.value().AsFFI() }, true }) : (diplomat::capi::OptionEnum_option{ {}, false }),
+  };
+}
+
+inline OptionInputStruct OptionInputStruct::FromFFI(diplomat::capi::OptionInputStruct c_struct) {
+  return OptionInputStruct {
+    /* .a = */ c_struct.a.is_ok ? std::optional(c_struct.a.ok) : std::nullopt,
```
Comment (sffc) [inline feature_tests/cpp/include/OptionInputStruct.hpp:36]: Question/Suggestion: Also down here; use `OptionU8::FromFFI(c_struct.a)` ?

File: macro/src/lib.rs [hunk 4]
```diff
@@ -324,13 +325,6 @@ fn gen_bridge(mut input: ItemMod) -> ItemMod {
             // never referenced. #[diplomat::transparent_convert] handles adding repr(transparent)
             // on its own
             if !info.opaque {
-                let copy = if !info.is_out {
-                    // Nothing stops FFI from copying, so we better make sure the struct is Copy.
-                    quote!(#[derive(Clone, Copy)])
-                } else {
-                    quote!()
-                };
```
Comment (sffc) [inline macro/src/lib.rs:332]: Question: Why this change?

**Review by sffc — APPROVED** [review]
Very straightforward

File: tool/templates/c/enum.h.jinja (hunk 1 above)
Comment (Manishearth) [reply tool/templates/c/enum.h.jinja:8]: There are potential safety issues around strict aliasing etc if it is not marked as a union. The field is allowed to be uninit, we mark it as such with a union.

File: macro/src/lib.rs (hunk 4 above)
Comment (Manishearth) [reply macro/src/lib.rs:332]: DiplomatOption cannot implement Copy, unfortunately, because it has a custom Drop impl. There's currently no way in Rust to have "Drop only for nontrivial fields" aside from the default drop glue.

The only thing requiring Copy was this sanity check. There are now Copy types allowed in input structs, so the sanity check is now incorrect and needed to be removed.

File: feature_tests/cpp/include/OptionInputStruct.hpp (hunk 2 above)
Comment (Manishearth) [reply feature_tests/cpp/include/OptionInputStruct.hpp:28]: I'm confused, there is no such method anywhere.

File: feature_tests/cpp/include/OptionInputStruct.hpp (hunk 3 above)
Comment (Manishearth) [reply feature_tests/cpp/include/OptionInputStruct.hpp:36]: Where is that defined?

File: macro/src/lib.rs (hunk 4 above)
Comment (robertbastian) [reply macro/src/lib.rs:332]: > There are now Copy types allowed in input structs, so the sanity check is now incorrect and needed to be removed.

What?

Comment (Manishearth) [reply macro/src/lib.rs:332]: sorry, non-Copy types. DiplomatOption isn't Copy

File: feature_tests/cpp/include/OptionInputStruct.hpp (hunk 2 above)
Comment (sffc) [reply feature_tests/cpp/include/OptionInputStruct.hpp:28]: My suggestion was that such a method should be created and called here.

File: feature_tests/cpp/include/OptionInputStruct.hpp (hunk 3 above)
Comment (sffc) [reply feature_tests/cpp/include/OptionInputStruct.hpp:36]: My suggestion was that such a method should be created and called here.

File: feature_tests/cpp/include/OptionInputStruct.hpp (hunk 2 above)
Comment (Manishearth) [reply feature_tests/cpp/include/OptionInputStruct.hpp:28]: This is a type from the C header: adding method to it would involve more C-CPP splits, or more duplicated code listing all of the primitives.  The code here seems rather simple, too, I don't think there's much value in trying to do it as a function.
